File: 2015/9.jl
using Combinatorics

function part1()
    f = open("instructions9a.txt", "r")
    lines = readlines(f)

    distances = Dict{String, Dict{String, Int32}}()

    for l in lines
        items = split(l, " ")
        d = parse(Int32, items[end])
        f = items[1]
        t = items[3]
        if haskey(distances, f)
            distances[f][t] = d
        else
            distances[f] = Dict(t => d)
        end

        if haskey(distances, t)
            distances[t][f] = d
        else
            distances[t] = Dict(f => d)
        end
    end 

    min_distance = 1e8
    max_distance = 0
    k = collect(keys(distances))
    for p in unique(permutations(k))
        dist = 0
        for i in 1:length(p)-1
            dist += distances[p[i]][p[i+1]]
        end
        if dist < min_distance
            min_distance = dist
        end
    
        if dist > max_distance
            max_distance = dist
        end
    end
    
    println("🎅 ", min_distance)
    println("🎄 ", max_distance)
end

part1()

File: 2015/9.py
import time, math

from itertools import permutations

from utils import readInput

def loadInput():
    lines = readInput("instructions9a.txt")
    distances = {}

    for d in lines:
        items = d.split(" ")
        distances.setdefault(items[0], {}).update({items[2]:int(items[4])})
        distances.setdefault(items[2], {}).update({items[0]:int(items[4])})
    return distances

def part1(distances):
    min_dist = math.inf
    for p in permutations(distances.keys()):
        dist = 0
        for i in range(len(p)-1):
            dist += distances[p[i]][p[i+1]]

        if dist < min_dist:
            min_dist = dist
            
    print (f"🎄 Part 1: {min_dist}")


def part2(distances):
    max_dist = 0
    for p in permutations(distances.keys()):
        dist = 0
        for i in range(len(p)-1):
            dist += distances[p[i]][p[i+1]]

        if dist > max_dist:
            max_dist = dist
            
    print (f"🎄🎅 Part 2: {max_dist}")

if __name__ == "__main__":
    title = "Day 9: All in a Single Night"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()
    
    t0 = time.time()
    part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
